include/deal.II/base/mpi.templates.h

```diff
@@ -324,6 +324,77 @@ namespace Utilities
 
 
 
+    template<typename T>
+    std::vector<T> gather(const MPI_Comm &comm,
+                          const T &object,
+                          const unsigned int root_process)
```

Comment:
I think this function should go into the `.h` file. We can't predict the type `T` for any reasonable subset of cases.

Reply:
Correct me if I'm wrong, but I thought just the function declaration is in the `.h` file and the definition is placed in the `.templates.h` file.
That is at least the case for the other wrappers in [include/deal.II/base/mpi.h](https://github.com/dealii/dealii/pull/5784/files/d9d834765800fa76bc3b95b0cd5cac4ff45bcdb1#diff-51e42c602d7e746a09658ab4fae843cc), where you can also find `Utilities::MPI::all_gather`.

Reply:
I think this falls under the last category of the discussion here: doc/doxygen/headers/instantiations.h

I tend to think that the `all_gather()` function has ended up in the `.templates.h` file by mistake. I'd accept a (separate) patch that moves it into the `.h` file, if you felt enterprising :-)

Reply:
@bangerth - I will have look where the other misplaced MPI functions are used.